This is Part 2 of a <a href="http://gettoknowdata.com/category/exploring-data/hotel-bookings/">mult-part series</a> exploring a dataset with data on hotel bookings.

I found the original hotel booking demand data at <a href="https://www.sciencedirect.com/science/article/pii/S2352340918315191" target="_blank">ScienceDirect</a>. It's also available in a slightly more processed form on <a href="https://www.kaggle.com/jessemostipak/hotel-booking-demand" target="_blank">Kaggle</a>.

This first block of code gets us to the point where we left off in <a href="http://gettoknowdata.com/hotel-bookings-part-1/" target="_blank">Part 1</a>.

In [1]:
########################################
######## START CODE FROM PART 1 ########
########################################

# Import pandas 
# so we can view the data 
# as a dataframe
import pandas as pd 

# Read the resort hotel data (H1.csv) 
# and city hotel data (H2.csv)
# into two separate dataframes 
# named resort_hotel and
# city_hotel  data, respecively
resort_hotel = pd.read_csv('H1.csv')
city_hotel = pd.read_csv('H2.csv')

# Change all of the NaN values
# in the Children column
# of the city hotel dataframe
# from NaN to 0
city_hotel['Children'].fillna(value=0,
                              inplace=True)

# Change the datatype 
# of the Children column
# in the city hotel dataframe
# to int64
city_hotel['Children'] = city_hotel['Children'].astype('int64')

# Add a Hotel column to 
# both the resort hotel dataframe
# and the city hotel dataframe
# and fill each with 'Resort' and
# 'City', respectively"
resort_hotel['Hotel'] = 'Resort'
city_hotel['Hotel'] = 'City'

# Combine (concatenate) the
# resort hotel dataframe and the
# city hotel dataframe
# into a single dataframe
# named resort_and_city_hotel
resort_and_city_hotel = pd.concat([resort_hotel,
                                   city_hotel],
                                  axis=0)

# Reset the index of the
# resort and city hotel dataframe
resort_and_city_hotel.reset_index(drop=True,
                                  inplace=True)

########################################
######## END CODE FROM PART 1 ##########
########################################

Let's take a look at the resort and city hotel dataframe.

In [2]:
display(resort_and_city_hotel)

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0                0       342             2015             July   
1                0       737             2015             July   
2                0         7             2015             July   
3                0        13             2015             July   
4                0        14             2015             July   
...            ...       ...              ...              ...   
119385           0        23             2017           August   
119386           0       102             2017           August   
119387           0        34             2017           August   
119388           0       109             2017           August   
119389           0       205             2017           August   

        ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                          27                      1                     0   
1                          27                      1                     0   
2                          27                      1                     0   
3                          27                      1                     0   
4                          27                      1                     0   
...                       ...                    ...                   ...   
119385                     35                     30                     2   
119386                     35                     31                     2   
119387                     35                     31                     2   
119388                     35                     31                     2   
119389                     35                     29                     2   

        StaysInWeekNights  Adults  Children  ...        Agent      Company  \
0                       0       2         0  ...         NULL         NULL   
1                       0       2         0  ...         NULL         NULL   
2                       1       1         0  ...         NULL         NULL   
3                       1       1         0  ...          304         NULL   
4                       2       2         0  ...          240         NULL   
...                   ...     ...       ...  ...          ...          ...   
119385                  5       2         0  ...          394         NULL   
119386                  5       3         0  ...            9         NULL   
119387                  5       2         0  ...            9         NULL   
119388                  5       2         0  ...           89         NULL   
119389                  7       2         0  ...            9         NULL   

       DaysInWaitingList CustomerType     ADR  RequiredCarParkingSpaces  \
0                      0    Transient    0.00                         0   
1                      0    Transient    0.00                         0   
2                      0    Transient   75.00                         0   
3                      0    Transient   75.00                         0   
4                      0    Transient   98.00                         0   
...                  ...          ...     ...                       ...   
119385                 0    Transient   96.14                         0   
119386                 0    Transient  225.43                         0   
119387                 0    Transient  157.71                         0   
119388                 0    Transient  104.40                         0   
119389                 0    Transient  151.20                         0   

        TotalOfSpecialRequests  ReservationStatus ReservationStatusDate  \
0                            0          Check-Out            2015-07-01   
1                            0          Check-Out            2015-07-01   
2                            0          Check-Out            2015-07-02   
3                            0          Check-Out            2015-07-02   
4                            1          Check-Out            2015

Okay. Great. We have what we expected: one dataframe with 119,390 records (rows) and 32 variables (columns).

The next thing I'd like to do is combine the ArrivalDateYear, ArrivalDateMonth, and ArrivalDateDayOfMonth variables into a single datetime variable. 

Let's do that now.

First, we'll import calendar and create a dictionary that maps the categorical names of the months to integers. Then we'll replace the names of the months in the ArrivalDateMonth variable with the integers stored in the dictionary. 

In [3]:
# Import calendar 
# so we can work with dates and times
import calendar

# Map month names to integers
months = {'January':1,
          'February':2,
          'March':3,
          'April':4,
          'May': 5,
          'June': 6,
          'July': 7,
          'August': 8,
          'September': 9,
          'October': 10,
          'November': 11,
          'December': 12}

# Replace the categorical month names
# in the ArrivalDateMonth variable
# with the integers 
# from the months dictionary
resort_and_city_hotel.ArrivalDateMonth = resort_and_city_hotel.ArrivalDateMonth.map(months)

Let's make sure we did this correctly.

Here are the first five rows of the dataframe.

In [4]:
resort_and_city_hotel.head()

IsCanceled  LeadTime  ArrivalDateYear  ArrivalDateMonth  \
0           0       342             2015                 7   
1           0       737             2015                 7   
2           0         7             2015                 7   
3           0        13             2015                 7   
4           0        14             2015                 7   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     0   
4                     27                      1                     0   

   StaysInWeekNights  Adults  Children  ...        Agent      Company  \
0                  0       2         0  ...         NULL         NULL   
1                  0       2         0  ...         NULL         NULL   
2                  1       1         0  ...         NULL         NULL   
3                  1       1         0  ...          304         NULL   
4                  2       2         0  ...          240         NULL   

  DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0                 0    Transient   0.0                         0   
1                 0    Transient   0.0                         0   
2                 0    Transient  75.0                         0   
3                 0    Transient  75.0                         0   
4                 0    Transient  98.0                         0   

   TotalOfSpecialRequests  ReservationStatus ReservationStatusDate   Hotel  
0                       0          Check-Out            2015-07-01  Resort  
1                       0          Check-Out            2015-07-01  Resort  
2                       0          Check-Out            2015-07-02  Resort  
3                       0          Check-Out            2015-07-02  Resort  
4                       1          Check-Out            2015-07-03  Resort  

[5 rows x 32 columns]

Looks good. We know from the last time we looked at the dataframe, the first five records were from July 2015, so all those 7s in the ArrivalDateMonth column represent July. 

Before we combine (concatenate) the ArrivalDateYear, ArrivalDateMonth, and ArrivalDateDayOfMonth variables into a single variable, we need to rename those columns to year, month, and day, respectively.

In [5]:
# Rename the ArrivalDateYear, 
# ArrivalDateMonth, and 
# ArrivalDateDayOfMonth variables
# to year, month, day (respectively) 
# so the date concat will work
resort_and_city_hotel.rename(columns=
                             {'ArrivalDateYear': 'year',
                              'ArrivalDateMonth': 'month',
                              'ArrivalDateDayOfMonth': 'day'},
                             inplace=True)

Let's check and make sure we did that correctly.

In [6]:
resort_and_city_hotel.head()

IsCanceled  LeadTime  year  month  ArrivalDateWeekNumber  day  \
0           0       342  2015      7                     27    1   
1           0       737  2015      7                     27    1   
2           0         7  2015      7                     27    1   
3           0        13  2015      7                     27    1   
4           0        14  2015      7                     27    1   

   StaysInWeekendNights  StaysInWeekNights  Adults  Children  ...  \
0                     0                  0       2         0  ...   
1                     0                  0       2         0  ...   
2                     0                  1       1         0  ...   
3                     0                  1       1         0  ...   
4                     0                  2       2         0  ...   

         Agent      Company DaysInWaitingList CustomerType   ADR  \
0         NULL         NULL                 0    Transient   0.0   
1         NULL         NULL                 0    Transient   0.0   
2         NULL         NULL                 0    Transient  75.0   
3          304         NULL                 0    Transient  75.0   
4          240         NULL                 0    Transient  98.0   

   RequiredCarParkingSpaces  TotalOfSpecialRequests  ReservationStatus  \
0                         0                       0          Check-Out   
1                         0                       0          Check-Out   
2                         0                       0          Check-Out   
3                         0                       0          Check-Out   
4                         0                       1          Check-Out   

  ReservationStatusDate   Hotel  
0            2015-07-01  Resort  
1            2015-07-01  Resort  
2            2015-07-02  Resort  
3            2015-07-02  Resort  
4            2015-07-03  Resort  

[5 rows x 32 columns]

Okay. Great. That's what we wanted. The ArrivalDateYear, ArrivalDateMonth, and ArrivalDateDayOfMonth columns have been renamed  to year, month, and day, respectively.

Now, let's create a new column named ArrivalDate that combines the (currently separate) year, month, and day columns into a single date with a datetime datatype.

In [7]:
# Add the new ArrivalDate column
# cast this new column as
# a datetime datatype
# and combine the the 
# year, month, and day columns 
# to create the (new) dates
resort_and_city_hotel['ArrivalDate'] = pd.to_datetime(resort_and_city_hotel[[
    'year', 
    'month', 
    'day']])

Check our work . . .

In [8]:
resort_and_city_hotel.head()

IsCanceled  LeadTime  year  month  ArrivalDateWeekNumber  day  \
0           0       342  2015      7                     27    1   
1           0       737  2015      7                     27    1   
2           0         7  2015      7                     27    1   
3           0        13  2015      7                     27    1   
4           0        14  2015      7                     27    1   

   StaysInWeekendNights  StaysInWeekNights  Adults  Children  ...  \
0                     0                  0       2         0  ...   
1                     0                  0       2         0  ...   
2                     0                  1       1         0  ...   
3                     0                  1       1         0  ...   
4                     0                  2       2         0  ...   

       Company DaysInWaitingList CustomerType   ADR RequiredCarParkingSpaces  \
0         NULL                 0    Transient   0.0                        0   
1         NULL                 0    Transient   0.0                        0   
2         NULL                 0    Transient  75.0                        0   
3         NULL                 0    Transient  75.0                        0   
4         NULL                 0    Transient  98.0                        0   

   TotalOfSpecialRequests  ReservationStatus  ReservationStatusDate   Hotel  \
0                       0          Check-Out             2015-07-01  Resort   
1                       0          Check-Out             2015-07-01  Resort   
2                       0          Check-Out             2015-07-02  Resort   
3                       0          Check-Out             2015-07-02  Resort   
4                       1          Check-Out             2015-07-03  Resort   

  ArrivalDate  
0  2015-07-01  
1  2015-07-01  
2  2015-07-01  
3  2015-07-01  
4  2015-07-01  

[5 rows x 33 columns]

Did you scroll all the way to the right?

The ArrivalDate column is currently the last column. Now, in the grand scheme of things, it doesn't matter where a column is in a dataframe. You'd still be able to run an analysis on a dataframe even if the columns don't appear in a logical order. Me? I like to put columns in an order that adds an additional layer of insight into the data. So, we'll reorder the columns later. When we do, I'll say a little more about why.

For now, though, let's drop the year, month, and day columns. We don't need them any more.

In [9]:
# Drop the year, month, and day columns
# from the resort and city hotel dataframe
resort_and_city_hotel.drop(['year',
                            'month',
                            'day'],
                           axis=1,
                           inplace=True)

And check our work . . .

In [10]:
resort_and_city_hotel.head()

IsCanceled  LeadTime  ArrivalDateWeekNumber  StaysInWeekendNights  \
0           0       342                     27                     0   
1           0       737                     27                     0   
2           0         7                     27                     0   
3           0        13                     27                     0   
4           0        14                     27                     0   

   StaysInWeekNights  Adults  Children  Babies       Meal Country  ...  \
0                  0       2         0       0  BB            PRT  ...   
1                  0       2         0       0  BB            PRT  ...   
2                  1       1         0       0  BB            GBR  ...   
3                  1       1         0       0  BB            GBR  ...   
4                  2       2         0       0  BB            GBR  ...   

       Company DaysInWaitingList  CustomerType   ADR  \
0         NULL                 0     Transient   0.0   
1         NULL                 0     Transient   0.0   
2         NULL                 0     Transient  75.0   
3         NULL                 0     Transient  75.0   
4         NULL                 0     Transient  98.0   

   RequiredCarParkingSpaces TotalOfSpecialRequests ReservationStatus  \
0                         0                      0         Check-Out   
1                         0                      0         Check-Out   
2                         0                      0         Check-Out   
3                         0                      0         Check-Out   
4                         0                      1         Check-Out   

   ReservationStatusDate   Hotel ArrivalDate  
0             2015-07-01  Resort  2015-07-01  
1             2015-07-01  Resort  2015-07-01  
2             2015-07-02  Resort  2015-07-01  
3             2015-07-02  Resort  2015-07-01  
4             2015-07-03  Resort  2015-07-01  

[5 rows x 30 columns]

Excellent. We're down to 30 variables. Always good to reduce the number of variables whenever possible&#8212;but only when it won't have a negative effect on the analysis you want to conduct. In this case, we reduced three columns that represented one date to one column that represents the same date. 

Speaking of reducing variables. Let's drop the ArrivalDateWeekNumber variable, too. We might want to know the week number when we start analyzing the data, but that's something we can calculate from the ArrivalDate variable. We don't need to store the number of the arrival week in the dataframe.

We can see from the first five rows of the dataframe that the second reservation had a lead time of 737 days and arrived on July 1, 2015&#8212;the 27th week of the year.

Here's a quick example of how to get the week number from a datetime (in case we ever need the week number).

In [11]:
# In the resort and city hotel dataframe,
# what is the week number
# of the reservation
# with a lead time equal to 737 days?
resort_and_city_hotel.loc[resort_and_city_hotel['LeadTime'] == 737,
                          'ArrivalDate'].dt.date.values[0].isocalendar()[1]

27

See? We don't need to store that value in the dataframe.

So, let's get rid of that column.

In [12]:
# Drop the ArrivalDateWeekNumber
# column from the dataframe
resort_and_city_hotel.drop(['ArrivalDateWeekNumber'],
                           axis=1,
                           inplace=True)

Check our work . . .

In [13]:
resort_and_city_hotel

IsCanceled  LeadTime  StaysInWeekendNights  StaysInWeekNights  Adults  \
0                0       342                     0                  0       2   
1                0       737                     0                  0       2   
2                0         7                     0                  1       1   
3                0        13                     0                  1       1   
4                0        14                     0                  2       2   
...            ...       ...                   ...                ...     ...   
119385           0        23                     2                  5       2   
119386           0       102                     2                  5       3   
119387           0        34                     2                  5       2   
119388           0       109                     2                  5       2   
119389           0       205                     2                  7       2   

        Children  Babies       Meal Country  MarketSegment  ...      Company  \
0              0       0  BB            PRT         Direct  ...         NULL   
1              0       0  BB            PRT         Direct  ...         NULL   
2              0       0  BB            GBR         Direct  ...         NULL   
3              0       0  BB            GBR      Corporate  ...         NULL   
4              0       0  BB            GBR      Online TA  ...         NULL   
...          ...     ...        ...     ...            ...  ...          ...   
119385         0       0  BB            BEL  Offline TA/TO  ...         NULL   
119386         0       0  BB            FRA      Online TA  ...         NULL   
119387         0       0  BB            DEU      Online TA  ...         NULL   
119388         0       0  BB            GBR      Online TA  ...         NULL   
119389         0       0  HB            DEU      Online TA  ...         NULL   

        DaysInWaitingList  CustomerType     ADR RequiredCarParkingSpaces  \
0                       0     Transient    0.00                        0   
1                       0     Transient    0.00                        0   
2                       0     Transient   75.00                        0   
3                       0     Transient   75.00                        0   
4                       0     Transient   98.00                        0   
...                   ...           ...     ...                      ...   
119385                  0     Transient   96.14                        0   
119386                  0     Transient  225.43                        0   
119387                  0     Transient  157.71                        0   
119388                  0     Transient  104.40                        0   
119389                  0     Transient  151.20                        0   

       TotalOfSpecialRequests  ReservationStatus ReservationStatusDate  \
0                           0          Check-Out            2015-07-01   
1                           0          Check-Out            2015-07-01   
2                           0          Check-Out            2015-07-02   
3                           0          Check-Out            2015-07-02   
4                           1          Check-Out            2015-07-03   
...                       ...                ...                   ...   
119385                      0          Check-Out            2017-09-06   
119386                      2          Check-Out            2017-09-07   
119387                      4          Check-Out            2017-09-07   
119388                      0          Check-Out            2017-09-07   
119389                      2          Check-Out            2017-09-07   

         Hotel ArrivalDate  
0       Resort  2015-07-01  
1       Resort  2015-07-01  
2       Resort  2015-07-01  
3       Resort  2015-07-01  
4       Resort  2015-07-01  
...        ...         ...  
119385    City  2017-08-30  
119386    City  2017-08-31  
119387    C

We're down to 29 variables.

We can also remove the StaysInWeekendNights and StaysInWeekNights variables&#8212;for the same reason we removed the ArrivalDateWeekNumber variable. This is something we can calculate based on other data in the dataframe, namely the ArrivalDate, ReservationStatusDate, and ReservationStatus columns. We can always use Python to make that calculation for us if we need it during analysis (maybe we'll do that in another post). 

For now, let's drop the StaysInWeekendNights and StaysInWeekNights columns.

In [14]:
# Drop the StaysInWeekendNights
# and StaysInWeekNights
# columns from the dataframe
resort_and_city_hotel.drop(['StaysInWeekendNights',
                            'StaysInWeekNights'],
                           axis=1,
                           inplace=True)

Check our work . . .

In [15]:
resort_and_city_hotel

IsCanceled  LeadTime  Adults  Children  Babies       Meal Country  \
0                0       342       2         0       0  BB            PRT   
1                0       737       2         0       0  BB            PRT   
2                0         7       1         0       0  BB            GBR   
3                0        13       1         0       0  BB            GBR   
4                0        14       2         0       0  BB            GBR   
...            ...       ...     ...       ...     ...        ...     ...   
119385           0        23       2         0       0  BB            BEL   
119386           0       102       3         0       0  BB            FRA   
119387           0        34       2         0       0  BB            DEU   
119388           0       109       2         0       0  BB            GBR   
119389           0       205       2         0       0  HB            DEU   

        MarketSegment DistributionChannel  IsRepeatedGuest  ...      Company  \
0              Direct              Direct                0  ...         NULL   
1              Direct              Direct                0  ...         NULL   
2              Direct              Direct                0  ...         NULL   
3           Corporate           Corporate                0  ...         NULL   
4           Online TA               TA/TO                0  ...         NULL   
...               ...                 ...              ...  ...          ...   
119385  Offline TA/TO               TA/TO                0  ...         NULL   
119386      Online TA               TA/TO                0  ...         NULL   
119387      Online TA               TA/TO                0  ...         NULL   
119388      Online TA               TA/TO                0  ...         NULL   
119389      Online TA               TA/TO                0  ...         NULL   

        DaysInWaitingList CustomerType     ADR  RequiredCarParkingSpaces  \
0                       0    Transient    0.00                         0   
1                       0    Transient    0.00                         0   
2                       0    Transient   75.00                         0   
3                       0    Transient   75.00                         0   
4                       0    Transient   98.00                         0   
...                   ...          ...     ...                       ...   
119385                  0    Transient   96.14                         0   
119386                  0    Transient  225.43                         0   
119387                  0    Transient  157.71                         0   
119388                  0    Transient  104.40                         0   
119389                  0    Transient  151.20                         0   

       TotalOfSpecialRequests ReservationStatus ReservationStatusDate   Hotel  \
0                           0         Check-Out            2015-07-01  Resort   
1                           0         Check-Out            2015-07-01  Resort   
2                           0         Check-Out            2015-07-02  Resort   
3                           0         Check-Out            2015-07-02  Resort   
4                           1         Check-Out            2015-07-03  Resort   
...                       ...               ...                   ...     ...   
119385                      0         Check-Out            2017-09-06    City   
119386                      2         Check-Out            2017-09-07    City   
119387                      4         Check-Out            2017-09-07    City   
119388                      0         Check-Out            2017-09-07    City   
119389                      2         Check-Out            2017-09-07    City   

       ArrivalDate  
0       2015-07-01  
1       2015-07-01  
2       2015-07-01  
3       2015-07-01  
4       2015-07-01  
...            ...  
119385  2017-08-30  
119386  2017-08-31  
119387  2017-08-31  
119388  2017-08-31  
11938

Looking good. We've eliminated 5 variables all without changing the nature of what this dataframe stores. 

Let's take another look at the info for our resort and city hotel dataframe.

In [16]:
resort_and_city_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 27 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   IsCanceled                   119390 non-null  int64         
 1   LeadTime                     119390 non-null  int64         
 2   Adults                       119390 non-null  int64         
 3   Children                     119390 non-null  int64         
 4   Babies                       119390 non-null  int64         
 5   Meal                         119390 non-null  object        
 6   Country                      118902 non-null  object        
 7   MarketSegment                119390 non-null  object        
 8   DistributionChannel          119390 non-null  object        
 9   IsRepeatedGuest              119390 non-null  int64         
 10  PreviousCancellations        119390 non-null  int64         
 11  PreviousBookingsNotCancele

Wait a minute. We have the right number of columns and column names, but look at the datatype for the ReservationStatusDate column. It's an object! We need to change that to datetime so if we ever decide to run time-series-based analysis on that column, it'll be treated like a date.

Let's change that datatype now.

In [17]:
# Change the ReservationStatusDate 
# to the datetime64[ns] datatype
resort_and_city_hotel['ReservationStatusDate'] = resort_and_city_hotel['ReservationStatusDate'].astype('datetime64[ns]')

Check our work . . .

In [18]:
resort_and_city_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 27 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   IsCanceled                   119390 non-null  int64         
 1   LeadTime                     119390 non-null  int64         
 2   Adults                       119390 non-null  int64         
 3   Children                     119390 non-null  int64         
 4   Babies                       119390 non-null  int64         
 5   Meal                         119390 non-null  object        
 6   Country                      118902 non-null  object        
 7   MarketSegment                119390 non-null  object        
 8   DistributionChannel          119390 non-null  object        
 9   IsRepeatedGuest              119390 non-null  int64         
 10  PreviousCancellations        119390 non-null  int64         
 11  PreviousBookingsNotCancele

Looking at the info again, I see there are a few more datatypes I want to change. 

The <a href="https://www.sciencedirect.com/science/article/pii/S2352340918315191#t0005" target="_blank">Variables Description</a> table in the <a href="https://www.sciencedirect.com/science/article/pii/S2352340918315191" target="_blank">ScienceDirect</a> article tells us that the following variables are *categorical* variables:

* IsCanceled
* IsRepeatedGuest

But if you look at the info, above, you'll see that these variables are both ints. Why are they ints if they're categorical variables? Because both of these variables are being stored as numerals&#8212;namely zeros and ones.

For the IsCanceled variable, 0 means the reservation *was not* canceled, and 1 means the reservation *was* canceled.

For the IsRepeatedGuest variable, 0 means the guest *is not* a repeat customer, and 1 means the guest *is* a repeat customer.

From a *statistics* perspective, these are indeed categorical variables. But when we look at how Python is storing them in the dataframe, they're being stored as ints. This feels like a disconnect to me. Why can't the *Python* datatypes align with the *statistical* datatypes?  

To me, if a variable implies a question that can be answered yes or no, then why not use "y" and "n" instead of 0 and 1? That would keep any *statistical* categorical variables stored as objects (instead of ints). (*Side note:* Maybe, depending on the character encoding, zeros and ones take up less memory than letters? I don't know. If that is the case, though, I suspect it really doesn't start to matter until you're working with massive datasets that are accessed millions or billions of times a day&#8212;like transactions in a large bank or at a large retailer.)

So, for both the IsCanceled and IsRepeatedGuest variables, let's change every 0 to an n and every 1 to a y.

But before we do that, let's get a count of the number of zeros and ones in each of those columns so we can quickly double check our work after we make the update.

In [19]:
# Of the values 
# in the IsCanceled column 
# of the resort_and_city_hotel dataframe,
# how many are zeros and how many are ones?
resort_and_city_hotel['IsCanceled'].value_counts()  

0    75166
1    44224
Name: IsCanceled, dtype: int64

So, now we know that, of the 119,390 reservations, 75,166 were *not* canceled, and 44,224 *were* canceled (which kind of sounds like a lot! But we'll think about that a little more when we analyze the data in another post).

And how many zeros and ones are there in the IsRepeatedGuest column?

In [20]:
# Of the values 
# in the IsRepeatedGuest column 
# of the resort_and_city_hotel dataframe,
# how many are zeros and how many are ones?
resort_and_city_hotel['IsRepeatedGuest'].value_counts()  

0    115580
1      3810
Name: IsRepeatedGuest, dtype: int64

Now we also know that, of the 119,390 reservations, 115,580 were *not* repeat customers, and 3,810 *were* repeat customers.

Okay. Good to know. Now let's change the zeros and ones to ns and ys.

In [21]:
# In the resort_and_city_hotel dataframe
# change every 0 to an n
# and every 1 to a y
# in the IsCanceled and
# IsRepeatedGuest columns
resort_and_city_hotel.loc[resort_and_city_hotel['IsCanceled'] == 0, 'IsCanceled'] = 'n'
resort_and_city_hotel.loc[resort_and_city_hotel['IsCanceled'] == 1, 'IsCanceled'] = 'y'
resort_and_city_hotel.loc[resort_and_city_hotel['IsRepeatedGuest'] == 0, 'IsRepeatedGuest'] = 'n'
resort_and_city_hotel.loc[resort_and_city_hotel['IsRepeatedGuest'] == 1, 'IsRepeatedGuest'] = 'y'

Now, if we did this correctly, we should get the same value counts, but with ns and ys instead of zeros and ones.

Let's check IsCanceled first . . .

In [22]:
resort_and_city_hotel['IsCanceled'].value_counts()  

n    75166
y    44224
Name: IsCanceled, dtype: int64

So far, so good. We get the same count of ns as we did of zeros. Ditto for the ones and ys.

Now let's check IsRepeatedGuest . . . 

In [23]:
resort_and_city_hotel['IsRepeatedGuest'].value_counts()  

n    115580
y      3810
Name: IsRepeatedGuest, dtype: int64

The IsRepeatedGuest counts for n and y match the counts for zeros and ones. 

We're good to go with changing these two categorical variables (IsCanceled and IsRepeatedGuest) from being stored as ints to being stored as objects. 

Let's double check by looking at the info.

In [24]:
resort_and_city_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 27 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   IsCanceled                   119390 non-null  object        
 1   LeadTime                     119390 non-null  int64         
 2   Adults                       119390 non-null  int64         
 3   Children                     119390 non-null  int64         
 4   Babies                       119390 non-null  int64         
 5   Meal                         119390 non-null  object        
 6   Country                      118902 non-null  object        
 7   MarketSegment                119390 non-null  object        
 8   DistributionChannel          119390 non-null  object        
 9   IsRepeatedGuest              119390 non-null  object        
 10  PreviousCancellations        119390 non-null  int64         
 11  PreviousBookingsNotCancele

Yup. This confirms it. Both of those columns are now objects. (*Side note:* I'm purposely keeping the objects as objects right now and not converting them to the category datatype. We might do that in another post.)

In fact, now every *Python* datatype aligns with the correct *statistical* datatype: every categorical (qualitative) variable is an object, and every quantitative variable is either an int, a float, or a datetime. 

In fact, ADR is the only variable that should be a float.

The last thing I'd like to do in this post is group and order the columns.

Do we need to do this? Probably not.

To me, though, it provides some perspective and deepens my understanding of the data. (*Side note:*  I realize this exercise is relatively simple when you only have about 30 variables to deal with. When you're working with datasets that have hundreds or thousands of variables, this could take far more time than it's worth.) 

In this case, we have a dataframe that represents reservations for hotels. Each record in the dataframe represents one reservation. 

So what is this dataframe representing? This dataframe is essentially capturing key data along the customer journey from making a reservation, to staying at the hotel, and leaving. Does it capture every data point in this process? No. But there is an implied *process* a customer goes through when making a reservation. 

Thinking about this data from the perspective of the customer going through a process helps me understand the data a little better.

So, let's take another look at the columns.

In [25]:
resort_and_city_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 27 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   IsCanceled                   119390 non-null  object        
 1   LeadTime                     119390 non-null  int64         
 2   Adults                       119390 non-null  int64         
 3   Children                     119390 non-null  int64         
 4   Babies                       119390 non-null  int64         
 5   Meal                         119390 non-null  object        
 6   Country                      118902 non-null  object        
 7   MarketSegment                119390 non-null  object        
 8   DistributionChannel          119390 non-null  object        
 9   IsRepeatedGuest              119390 non-null  object        
 10  PreviousCancellations        119390 non-null  int64         
 11  PreviousBookingsNotCancele

There's no right or wrong answer here. If you come up with different groupings, that's perfectly fine.

I see three large groups of variables (main bullets) and some subgroups (sub-bullets with questions and variable names): 
* **Things the Customer Would Identify or Choose**
    * How many adults, children, and babies are in the customer's party?
        * Adults                     
        * Children
        * Babies
    * What type of meal package has the customer selected?
        * Meal
    * What kind of room has the customer booked?
        * ReservedRoomType 
    * How many parking spaces does the customer need?
        * RequiredCarParkingSpaces 
    * What is the primary customer's country of origin?
        * Country
    * On what date will the customer arrive?
        * ArrivalDate
    * Did a travel agent make the reservation on behalf of the customer?
        * Agent
    * Did the primary customer's company make the reservation on behalf of the guest?
        * Company
* **Things the Business Measures about the Customer**
    * What type of customer is making this reservation?
        * CustomerType
    * Has this customer stayed with us before?
        * IsRepeatedGuest
    * If the customer is a repeat customer, have they ever canceled a reservation?
        * PreviousCancellations
    * If the customer is a repeat customer, how many of their previous bookings have they **not** canceled?
        * PreviousBookingsNotCanceled
* **Things the Business Measures about the Reservation**
    * For which hotel did the customer make this reservation?
        * Hotel
    * Has the customer canceled this reservation?
        * IsCanceled
    * How many times has the customer changed the booking since they initiated the reservation?
        * BookingChanges
    * How many days was the reservation in the waiting list before being confirmed to the customer?
        * DaysInWaitingList
    * How many special requests has the customer made for this reservation?
        * TotalOfSpecialRequests
    * What type of deposit did the customer put down on the reservation?
        * DepositType
    * What market segment did this reservation originate from?
        * MarketSegment
    * What distribution channel did this reservation originate from?
        * DistributionChannel
    * When the customer part arrived, what room type did we assign them?
        * AssignedRoomType
    * What was the ADR for this reservation?
        * ADR
    * How many days was it from the time the customer made the reservation to the time they either arrived for their stay or canceled their reservation?
        * LeadTime
    * What is the current status of this reservation?
        * ReservationStatus
    * On what date was the reservation status updated?
        * ReservationStatusDate

Again, this is just a way for me to get to know the data a little better before I start analyzing it. And, as I mentioned in my side note above, this really wouldn't be feasible for massive datasets with hundreds or thousands or even millions of variables.

All that being said, let's reorder the columns in the order implied by the bullet list above.

In [26]:
# Reorder the columns in the
# resort_and_city_hotel dataframe
resort_and_city_hotel = resort_and_city_hotel[
    ['Adults',
     'Children',
     'Babies',
     'Meal',
     'ReservedRoomType',
     'RequiredCarParkingSpaces',
     'Country',
     'ArrivalDate',
     'Agent',
     'Company',
     'CustomerType',
     'IsRepeatedGuest',
     'PreviousCancellations',
     'PreviousBookingsNotCanceled',
     'Hotel',
     'IsCanceled',
     'BookingChanges',
     'DaysInWaitingList',
     'TotalOfSpecialRequests',
     'DepositType',
     'MarketSegment',
     'DistributionChannel',
     'AssignedRoomType',
     'ADR',
     'LeadTime',
     'ReservationStatus',
     'ReservationStatusDate']]

And check our work . . .

In [27]:
resort_and_city_hotel

Adults  Children  Babies       Meal  ReservedRoomType  \
0            2         0       0  BB         C                  
1            2         0       0  BB         C                  
2            1         0       0  BB         A                  
3            1         0       0  BB         A                  
4            2         0       0  BB         A                  
...        ...       ...     ...        ...               ...   
119385       2         0       0  BB         A                  
119386       3         0       0  BB         E                  
119387       2         0       0  BB         D                  
119388       2         0       0  BB         A                  
119389       2         0       0  HB         A                  

        RequiredCarParkingSpaces Country ArrivalDate        Agent  \
0                              0     PRT  2015-07-01         NULL   
1                              0     PRT  2015-07-01         NULL   
2                              0     GBR  2015-07-01         NULL   
3                              0     GBR  2015-07-01          304   
4                              0     GBR  2015-07-01          240   
...                          ...     ...         ...          ...   
119385                         0     BEL  2017-08-30          394   
119386                         0     FRA  2017-08-31            9   
119387                         0     DEU  2017-08-31            9   
119388                         0     GBR  2017-08-31           89   
119389                         0     DEU  2017-08-29            9   

            Company  ... DaysInWaitingList TotalOfSpecialRequests  \
0              NULL  ...                 0                      0   
1              NULL  ...                 0                      0   
2              NULL  ...                 0                      0   
3              NULL  ...                 0                      0   
4              NULL  ...                 0                      1   
...             ...  ...               ...                    ...   
119385         NULL  ...                 0                      0   
119386         NULL  ...                 0                      2   
119387         NULL  ...                 0                      4   
119388         NULL  ...                 0                      0   
119389         NULL  ...                 0                      2   

            DepositType  MarketSegment DistributionChannel  AssignedRoomType  \
0       No Deposit              Direct              Direct  C                  
1       No Deposit              Direct              Direct  C                  
2       No Deposit              Direct              Direct  C                  
3       No Deposit           Corporate           Corporate  A                  
4       No Deposit           Online TA               TA/TO  A                  
...                 ...            ...                 ...               ...   
119385  No Deposit       Offline TA/TO               TA/TO  A                  
119386  No Deposit           Online TA               TA/TO  E                  
119387  No Deposit           Online TA               TA/TO  D                  
119388  No Deposit           Online TA               TA/TO  A                  
119389  No Deposit           Online TA               TA/TO  A                  

           ADR  LeadTime  ReservationStatus ReservationStatusDate  
0         0.00       342          Check-Out            2015-07-01  
1         0.00       737          Check-Out            2015-07-01  
2        75.00         7          Check-Out            2015-07-02  
3        75.00        13          Check-Out            2015-07-02  
4        98.00        14          Check-Out            2015-07-03  
...        ...       ...                ...                   ...  
119385   96.14        23          Check-Out            2017-09-06  
119386  225.43       102          Check-Out            2017-09-0

Okay! Looking good. That's it for this post. 

We now have our dataframe in pretty good shape, ready for analysis.

In upcoming parts of this series, we'll start analyzing the data to see what insights we can generate.

<a href="http://gettoknowdata.com/exploring-data-hotel-bookings-part-3/">Continue to Part 3</a>.

<a href="http://gettoknowdata.com/category/exploring-data/hotel-bookings/">All posts in the Hotel Bookings series</a>.